starknet-replay/src/storage/rpc/transaction.rs

```diff
@@ -1,102 +1,299 @@
-    Transaction,
+    InvokeTransactionV0,
+    InvokeTransactionV1,
+    InvokeTransactionV3,
+    L1HandlerTransaction,
+    PaymasterData,
+    Resource,
+    ResourceBounds,
+    Tip,
+    Transaction as StarknetApiTransaction,
+    TransactionSignature,
+    TransactionVersion,
 };
+use starknet_core::types::Transaction as StarknetCoreTransaction;
 
 use crate::error::DatabaseError;
 
-/// This function returns a [`starknet_api::transaction::Transaction`]
-/// from a serialised transaction.
+/// This function converts [`starknet_core::types::Transaction`] into
+/// [`starknet_api::transaction::Transaction`].
 ///
 /// # Arguments
 ///
-/// - `transaction`: the transaction in JSON format. It shall be of the format
-///   returned by `starknet_getBlockWithReceipts`.
+/// - `transaction`: The transaction object.
 ///
 /// # Errors
 ///
-/// Returns [`Err`] if the JSON data is not in the correct format.
+/// Returns [`Err`] if `transaction` contains invalid numbers that can't be
+/// translated to a [`starknet_api::transaction::Transaction`] object.
 #[allow(clippy::too_many_lines)] // Added because there is a lot of repetition.
-pub fn deserialize_transaction_json(
-    transaction: &serde_json::Value,
-) -> Result<Transaction, DatabaseError> {
-    let mut transaction = transaction.clone();
-    if let Some(resource_bounds) = transaction.get_mut("resource_bounds") {
-        if let Some(l1_gas) = resource_bounds.get_mut("l1_gas") {
-            resource_bounds["L1_GAS"] = l1_gas.clone();
-            resource_bounds
-                .as_object_mut()
-                .ok_or(DatabaseError::Unknown(
-                    "Failed to serialise transaction as object.".to_string(),
-                ))?
-                .remove("l1_gas");
-        }
-        if let Some(l2_gas) = resource_bounds.get_mut("l2_gas") {
-            resource_bounds["L2_GAS"] = l2_gas.clone();
-            resource_bounds
-                .as_object_mut()
-                .ok_or(DatabaseError::Unknown(
-                    "Failed to serialise transaction as object.".to_string(),
-                ))?
-                .remove("l2_gas");
-        }
-    }
-
-    let tx_type: String = serde_json::from_value(transaction["type"].clone())?;
-    let tx_version: String = serde_json::from_value(transaction["version"].clone())?;
-
-    match tx_type.as_str() {
-        "INVOKE" => match tx_version.as_str() {
-            "0x0" => Ok(Transaction::Invoke(InvokeTransaction::V0(
-                serde_json::from_value(transaction)?,
-            ))),
-            "0x1" => Ok(Transaction::Invoke(InvokeTransaction::V1(
-                serde_json::from_value(transaction)?,
-            ))),
-            "0x3" => Ok(Transaction::Invoke(InvokeTransaction::V3(
-                serde_json::from_value(transaction)?,
-            ))),
-            x => Err(DatabaseError::Unknown(format!(
-                "unimplemented invoke version: {x}"
-            ))),
+pub fn convert_transaction(
```

Comment:
Okay this method is absolutely mammoth. It needs to be split up, and thankfully the wrapped `tx` in most cases appears to be a different type that could be forwarded to another method.